File: snippets/fsharp/System/AppDomain/DoCallBack/docallback_byval.fs
module docallbackbyval

// <Snippet2>
open System

[<Serializable>]
type PingPong() as this =
    [<DefaultValue>]
    val mutable greetings: string
    do  
        this.greetings <- "PING!"

    member _.MyCallBack() =
        let mutable name = AppDomain.CurrentDomain.FriendlyName

        if name = AppDomain.CurrentDomain.SetupInformation.ApplicationName then
            name <- "defaultDomain"
        printfn $"{this.greetings} from {name}"

let otherDomain = AppDomain.CreateDomain "otherDomain"

let pp = PingPong()
pp.MyCallBack()
pp.greetings <- "PONG!"
otherDomain.DoCallBack(CrossAppDomainDelegate pp.MyCallBack)

// Output:
//   PING! from defaultDomain
//   PONG! from otherDomain
// </Snippet2>

File: snippets/csharp/System/AppDomain/DoCallBack/docallback_byval.cs
using System;

// <Snippet2>
[Serializable]
public class CallbackByValSnippet
{
    private string greetings = "PING!";

    public static void Main()
    {
        AppDomain otherDomain = AppDomain.CreateDomain("otherDomain");

        CallbackByValSnippet pp = new CallbackByValSnippet();
        pp.MyCallBack();
        pp.greetings = "PONG!";
        otherDomain.DoCallBack(new CrossAppDomainDelegate(pp.MyCallBack));

        // Output:
        //   PING! from defaultDomain
        //   PONG! from otherDomain
    }

    public void MyCallBack()
    {
        string name = AppDomain.CurrentDomain.FriendlyName;

        if (name == AppDomain.CurrentDomain.SetupInformation.ApplicationName)
        {
            name = "defaultDomain";
        }
        Console.WriteLine(greetings + " from " + name);
    }
}
// </Snippet2>
